### DISCUSSION TITLE
Cell disabling at notebook startup

### DISCUSSION THREAD

**Discussion by lungben** [fonsp/Pluto.jl#1205] [Pluto development] [answered]
Currently, the cell disabling #985 is only in the current Pluto session, but not persistent in the notebook.jl file. Especially, when opening a notebook all cells are active, which may lead to long execution times.

It should be possible in Pluto to mark cells which should be disabled when starting the notebook. This information is then stored in the notebook file.

Backend:
* Store the information if a cell is disabled as startup as a comment in the first line of the cell, e.g. 
`# 🛑 disabled at startup 🛑`
* The disabled code still stays active in the notebook.jl file itself so that it is executed if the notebook.jl file is run as a standard Julia file. Cell disabling is a feature for improving interactive notebook experience.

Frontend:
* For enabling / disabling the cell at startup, additional entries in the cell context menu (the same as for disabling cells in the current session) should be added.
* How do we show if a cell is disabled at notebook start?
  * Add a new UI element for it, e.g. 🛑 at the cell?
  * Just show the comment `# 🛑 disabled at startup 🛑` as first line in the cell itself (simplest solution)?

Happy to discuss!

**Comment by ederag** [comment, marked as answer]
Persistent disabling would be great indeed.

> The disabled code still stays active in the notebook.jl file itself
> so that it is executed if the notebook.jl file is run as a standard Julia file.

Just to fuel the discussion, I'd prefer the opposite.
Disabled cells can become obsolete, as they are not executed by Pluto.
So to me they are in the limbo, and should not be executed in the repl.
(Only cells that actually run in Pluto would be run when the file is included in the repl).

Technically, disabled cell code could be wrapped in a [multi-line comment](https://docs.julialang.org/en/v1/base/punctuation/), as
```julia
# ╔═╡ 45fab7a0-9076-11eb-284b-9b712e6f3d46
#=╠═ Cell manually disabled in Pluto (or "Cell disabled by Pluto because it depends on a disabled cell")
...
=#
```
The documentation says multi-line comments are nestable, 
so it should work even with other comments in the cell code.

**Reply by fonsp** [reply]
I was also thinking that "Cell disabling" should be codified as commenting-out-the-cell! That fits the ideal that Pluto is "making existing Julia features interactive", instead of adding our own.

Just using `#= ...cell code... =#` could lead to edge cases that break existing notebooks, so  I really like @ederag's suggestion of using a special first line, clever!

**Reply by lungben** [reply]
When we comment out the cells explicitly disabled for the next startup, we also need to comment out all their downstream dependencies (but with a different header line). Otherwise, the notebook file would not run stand-alone.

Technically, this would imply:
1. When Pluto opens a notebook, the cells marked as disabled on startup will have `Cell.running_disabled=true`, but not the indirectly disabled ones.
2. `Cell.code` should always contain the non-commented-out code, because cells can be activated intereactively in Pluto without removing the "disabled at startup" property. Thus, we need the dependency information of these cells.
3. Effectively, in each cell the following information is stored:

```
# current information - is cell disabled in the current Pluto session?
running_disabled::Bool=false
depends_on_disabled_cells::Bool=false
# new information - is cell directly or indirectly disabled at startup
running_disabled_on_startup::Bool=false
depends_on_disabled_cells_on_startup::Bool=false
```
4. On each notebook change, both the cell disabling information in the current Pluto session and on startup is updated independently (but with the same logic). The latter is directly saved to the notebook.jl file.
5. A new GUI element / effect showing `running_disabled_on_startup` for each cell, maybe also `depends_on_disabled_cells_on_startup`.

But in general this sounds feasible!

**Reply by ederag** [reply]
Any reason to make startup special ?
To me the dichotomy was between 

- "manually disabled" (I explicitly tell to disable a cell)
- "disabled as dependency"

Cell code is commented out in the saved file, together with something to explain why
(manually disabled, disabled as a dependency to a disabled cell, never executed, ...)

##### When the file is included in the REPL
Commented out cells do not execute, as in the last Pluto session.

##### When the file is reopened in Pluto
Pluto automatically flags cells that should be disabled, based on the user intention (the "manually disabled" cells).

**Reply by lungben** [reply]
> Any reason to make startup special ?

I was thinking about the case of a very long running cell which should be disabled when you open the notebook. In the Pluto session itself you can enable it again to execute it. In case of a crash (or whatever other interruption) the cell should be disabled again at the next startup.
The drawback of this is that it makes the whole logic in Pluto more complex because you have to track 2 dependency trees - one for the startup and one for the current session. 
What do you think?

**Reply by fonsp** [reply]
I don't think having two separate disable settings is a good idea, it is almost guaranteed not to make sense for new users.

**Reply by lungben** [reply]
OK - that also simplifies the implementation (no frontend changes required).

**Reply by lungben** [reply]
Got it running #1209 - please take a look at it

**Reply by lungben** [reply]
Implementation finished.
